components/org.wso2.carbon.security.sts.common/src/main/java/org/wso2/carbon/identity/sts/common/util/ServicePasswordCallbackHandler.java

```diff
@@ -381,44 +379,14 @@ private String getPrivateKeyPassword(String username) throws IOException, Except
 
         String password = null;
         int tenantId = ((UserRegistry) registry).getTenantId();
-        UserRegistry govRegistry = SecurityServiceHolder.getRegistryService().
-                getGovernanceSystemRegistry(tenantId);
         try {
             KeyStoreManager keyMan = KeyStoreManager.getInstance(tenantId);
-            if (govRegistry.resourceExists(SecurityConstants.KEY_STORES)) {
-                Collection collection = (Collection) govRegistry.get(SecurityConstants.KEY_STORES);
-                String[] ks = collection.getChildren();
-
-                for (int i = 0; i < ks.length; i++) {
-
-                    String fullname = ks[i];
-                    //get the primary keystore, only if it is super tenant.
-                    if (tenantId == MultitenantConstants.SUPER_TENANT_ID && fullname
-                            .equals(RegistryResources.SecurityManagement.PRIMARY_KEYSTORE_PHANTOM_RESOURCE)) {
-                        KeyStore store = keyMan.getPrimaryKeyStore();
-                        if (store.containsAlias(username)) {
-                            password = keyMan.getPrimaryPrivateKeyPasssword();
-                            break;
-                        }
-                    } else {
-                        String name = fullname.substring(fullname.lastIndexOf("/") + 1);
-                        KeyStore store = null;
-                        //Not all the keystores encrypted using primary keystore password. So, some of the keystores will fail while loading
-                        store = keyMan.getKeyStore(name);
-                        if (log.isDebugEnabled()) {
-                            log.debug("Load the keystore " + name);
-                        }
-                        if (store != null && store.containsAlias(username)) {
-                            Resource resource = (Resource) govRegistry.get(ks[i]);
-                            CryptoUtil cryptoUtil = CryptoUtil.getDefaultCryptoUtil();
-                            String encryptedPassword = resource
-                                    .getProperty(SecurityConstants.PROP_PRIVATE_KEY_PASS);
-                            password = new String(cryptoUtil
-                                    .base64DecodeAndDecrypt(encryptedPassword));
-                            break;
-                        }
-                    }
-
+            KeyStoreMetadata[] keyStoreMetadataArray = keyMan.getKeyStoresMetadata(SUPER_TENANT_ID == tenantId);
+            for (KeyStoreMetadata keyStoreMetadata : keyStoreMetadataArray) {
+                KeyStore keyStore = keyMan.getKeyStore(keyStoreMetadata.getKeyStoreName());
+                if (keyStore.containsAlias(username)) {
+                    password = String.valueOf(keyMan.getPrivateKeyPassword(keyStoreMetadata.getKeyStoreName()));
```

Comment:
Shouldn't this method return a chat[]?

Reply:
which method?

Reply:
If you are referring the `ServicePasswordCallbackHandler.getPrivateKeyPassword()` method, we can technically change the signature to return a char[], but its have little to no value, since `WSPasswordCallback.setPassword()` anyway requires a string, so we have to convert the char[] to string right after the returning from the `ServicePasswordCallbackHandler.getPrivateKeyPassword()` method..

[1] https://github.com/wso2-extensions/identity-inbound-auth-sts/blob/bc3702957e197f9bd0d68740d79bb6dfccab08af/components/org.wso2.carbon.security.sts.common/src/main/java/org/wso2/carbon/identity/sts/common/util/ServicePasswordCallbackHandler.java#L378